### PR TITLE
fix: [CDS-3897]: Support for Canary Rollback

### PR DESCRIPTION
**Pull request by TarunUba1**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/36264)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 2 Issues 
* Code issue 2 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/36264)

File: 400-rest/src/main/java/software/wings/sm/states/k8s/K8sDelete.java [hunk 1]
```diff
@@ -118,7 +119,13 @@ private ExecutionResponse executeWithoutManifest(ExecutionContext context) {
...
       Activity activity = createActivity(context);
 
-      K8sTaskParameters k8sTaskParameters =
+      K8sCanaryDeleteServiceElement k8sCanaryDeleteServiceElement = fetchK8sCanaryDeleteServiceElement(context);
+      K8sTaskParameters k8sTaskParameters;
+      boolean wasCanaryDeployedPrev = false;
+      if (k8sCanaryDeleteServiceElement != null && k8sCanaryDeleteServiceElement.getPreviousDeployedK8sCanary()) {
+        wasCanaryDeployedPrev = true;
+      }
+      k8sTaskParameters =
```
Comment (namantalaycha10) [inline 400-rest/src/main/java/software/wings/sm/states/k8s/K8sDelete.java:128]: Move this to a separate method , as Canary related logic is looking odd in K8sDelete  , only read here the boolean value response of that method

File: 260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java [hunk 2]
```diff
@@ -241,4 +269,36 @@ List<KubernetesResourceId> fetchAllCreatedResourceIdsForDeletion(
       K8sDeleteTaskParameters k8sDeleteTaskParameters, ExecutionLogCallback executionLogCallback) throws IOException {
     return k8sTaskHelper.getResourceIdsForDeletion(k8sDeleteTaskParameters, kubernetesConfig, executionLogCallback);
   }
+  private List<KubernetesResourceId> getCanaryResourceIdsFromReleaseHistory(String releaseName, LogCallback logCallback)
```
Comment (namantalaycha10) [inline 260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java:272]: Just if it is possible to reuse the method already written for the NG

File: 260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java [hunk 3]
```diff
@@ -159,6 +168,25 @@ private boolean init(K8sDeleteTaskParameters k8sDeleteTaskParameters, K8sDelegat
       kubernetesConfig =
           containerDeploymentDelegateHelper.getKubernetesConfig(k8sDeleteTaskParameters.getK8sClusterConfig(), false);
 
+      String resource_check = k8sDeleteTaskParameters.getResources();
+
+      if (k8sDeleteTaskParameters.isK8sPrevCanaryDeploy() == true && resource_check.contains("${k8s.canaryWorkload}")) {
```
Comment (namantalaycha10) [inline 260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java:173]: As isK8sPrevCanaryDeploy is a boolean no need to add "== true".

**Review by namantalaycha10 — CHANGES_REQUESTED** [review]

File: 260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java (hunk 2 above)
Comment (TarunUba1) [reply 260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java:272]: done

File: 260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java (hunk 3 above)
Comment (TarunUba1) [reply 260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java]: Done

**Review by shubham-in10se — CHANGES_REQUESTED** [review]
> Reviewed 6 of 8 files at r1, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36264)* status: 2 change requests. 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java` line 171 at r3](https://reviewable.io/reviews/harness/harness-core/36264#-N9BeOctDtIbNKHx0CoS:-N9BeOctDtIbNKHx0CoT:b2lps9n) ([raw file](https://github.com/harness/harness-core/blob/a085b968f96a6da4f8c4c1f7bc78ff7d569c349d/260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java#L171)):*
> ```Java
>           containerDeploymentDelegateHelper.getKubernetesConfig(k8sDeleteTaskParameters.getK8sClusterConfig(), false);
> 
>       String resource_check = k8sDeleteTaskParameters.getResources();
> ```

kindly use camel case naming conventions for java variables.
___
*[`260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java` line 171 at r3](https://reviewable.io/reviews/harness/harness-core/36264#-N9BeZ_b6R78es1shfbD:-N9BeZ_b6R78es1shfbE:bdmsp1y) ([raw file](https://github.com/harness/harness-core/blob/a085b968f96a6da4f8c4c1f7bc78ff7d569c349d/260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java#L171)):*
> ```Java
>           containerDeploymentDelegateHelper.getKubernetesConfig(k8sDeleteTaskParameters.getK8sClusterConfig(), false);
> 
>       String resource_check = k8sDeleteTaskParameters.getResources();
> ```

rename this variable. 
The variable should indicate some context for the value(s) it holds. In this case, this variable holds the name/reference of the K8s resource to delete.
___
*[`260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java` line 171 at r3](https://reviewable.io/reviews/harness/harness-core/36264#-N9BtU9zAb2frbuNeWYq:-N9BtUA--qJOlzrw8WL3:b-1yebzn) ([raw file](https://github.com/harness/harness-core/blob/a085b968f96a6da4f8c4c1f7bc78ff7d569c349d/260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java#L171)):*
> ```Java
>           containerDeploymentDelegateHelper.getKubernetesConfig(k8sDeleteTaskParameters.getK8sClusterConfig(), false);
> 
>       String resource_check = k8sDeleteTaskParameters.getResources();
> ```

this can be null as well.. ??
___
*[`260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java` line 173 at r3](https://reviewable.io/reviews/harness/harness-core/36264#-N9Bezlc0FYOiy0yAjgC:-N9Bezld822BuN1qzYu-:b-mfzf8e) ([raw file](https://github.com/harness/harness-core/blob/a085b968f96a6da4f8c4c1f7bc78ff7d569c349d/260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java#L173)):*
> ```Java
>       String resource_check = k8sDeleteTaskParameters.getResources();
> 
>       if (k8sDeleteTaskParameters.isK8sPrevCanaryDeploy() && resource_check.contains("${k8s.canaryWorkload}")) {
> ```

rename this new task param.
better name could be 'isK8sCanaryDelete', since we are checking if this delete is part of canary deployment, hence if this delete is canary delete.



_Code quote:_
```Java
k8sDeleteTaskParameters.isK8sPrevCanaryDeploy()
```
___
*[`260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java` line 183 at r3](https://reviewable.io/reviews/harness/harness-core/36264#-N9BtEJK2H7PsGE79mpy:-N9BtEJK2H7PsGE79mpz:b-l4q0fg) ([raw file](https://github.com/harness/harness-core/blob/a085b968f96a6da4f8c4c1f7bc78ff7d569c349d/260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java#L183)):*
> ```Java
> 
>         if (isEmpty(resourceIdsToDelete)) {
>           executionLogCallback.saveExecutionLog(color("No canary workloads to be deleted. Skipping.\n", White, Bold));
> ```

this can be an 'else' branch for the above 'if'



_Code quote:_
```Java
        if (isEmpty(resourceIdsToDelete)) {
          executionLogCallback.saveExecutionLog(color("No canary workloads to be deleted. Skipping.\n", White, Bold));
          executionLogCallback.saveExecutionLog("\nDone.", INFO, SUCCESS);
          return true;
        }
```
___
*[`260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java` line 185 at r3](https://reviewable.io/reviews/harness/harness-core/36264#-N9BtJG_4vWRI3D6ZAVl:-N9BtJG_4vWRI3D6ZAVm:blrbyv6) ([raw file](https://github.com/harness/harness-core/blob/a085b968f96a6da4f8c4c1f7bc78ff7d569c349d/260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java#L185)):*
> ```Java
>           executionLogCallback.saveExecutionLog(color("No canary workloads to be deleted. Skipping.\n", White, Bold));
>           executionLogCallback.saveExecutionLog("\nDone.", INFO, SUCCESS);
>           return true;
> ```

we are always returning 'true' from this 'if' block. We can remove the redundant return

**Review by TarunUba1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36264)* status: 5 file(s) to review. 2 change requests. 9. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java` line 171 at r3](https://reviewable.io/reviews/harness/harness-core/36264#-N9BeOctDtIbNKHx0CoS:-N9FsapQ1CS_viOp5nRy:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a085b968f96a6da4f8c4c1f7bc78ff7d569c349d/260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java#L171)):*
<details><summary><i>Previously, shubham-in10se (Shubham Maheshwari) wrote…</i></summary><blockquote>

kindly use camel case naming conventions for java variables.
</blockquote></details>

Done.
___
*[`260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java` line 171 at r3](https://reviewable.io/reviews/harness/harness-core/36264#-N9BeZ_b6R78es1shfbD:-N9FscTK5oPSF8d6Pt0Q:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a085b968f96a6da4f8c4c1f7bc78ff7d569c349d/260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java#L171)):*
<details><summary><i>Previously, shubham-in10se (Shubham Maheshwari) wrote…</i></summary><blockquote>

rename this variable. 
The variable should indicate some context for the value(s) it holds. In this case, this variable holds the name/reference of the K8s resource to delete.
</blockquote></details>

Done.
___
*[`260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java` line 171 at r3](https://reviewable.io/reviews/harness/harness-core/36264#-N9BtU9zAb2frbuNeWYq:-N9Fscnx815W5s5vdQAp:btzxet8) ([raw file](https://github.com/harness/harness-core/blob/a085b968f96a6da4f8c4c1f7bc78ff7d569c349d/260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java#L171)):*
<details><summary><i>Previously, shubham-in10se (Shubham Maheshwari) wrote…</i></summary><blockquote>

this can be null as well.. ??
</blockquote></details>

This case is already handled by existing method
___
*[`260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java` line 173 at r3](https://reviewable.io/reviews/harness/harness-core/36264#-N9Bezlc0FYOiy0yAjgC:-N9Fsk1h51dSa403-OGp:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a085b968f96a6da4f8c4c1f7bc78ff7d569c349d/260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java#L173)):*
<details><summary><i>Previously, shubham-in10se (Shubham Maheshwari) wrote…</i></summary><blockquote>

rename this new task param.
better name could be 'isK8sCanaryDelete', since we are checking if this delete is part of canary deployment, hence if this delete is canary delete.
</blockquote></details>

Done.
___
*[`260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java` line 183 at r3](https://reviewable.io/reviews/harness/harness-core/36264#-N9BtEJK2H7PsGE79mpy:-N9FslZ672SX85lSr4Iw:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a085b968f96a6da4f8c4c1f7bc78ff7d569c349d/260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java#L183)):*
<details><summary><i>Previously, shubham-in10se (Shubham Maheshwari) wrote…</i></summary><blockquote>

this can be an 'else' branch for the above 'if'
</blockquote></details>

Done.
___
*[`260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java` line 185 at r3](https://reviewable.io/reviews/harness/harness-core/36264#-N9BtJG_4vWRI3D6ZAVl:-N9FsmsD02KH4cIZLxHT:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a085b968f96a6da4f8c4c1f7bc78ff7d569c349d/260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java#L185)):*
<details><summary><i>Previously, shubham-in10se (Shubham Maheshwari) wrote…</i></summary><blockquote>

we are always returning 'true' from this 'if' block. We can remove the redundant return
</blockquote></details>

Done.
___
*[`400-rest/src/main/java/software/wings/sm/states/k8s/K8sDelete.java` line 128 at r2](https://reviewable.io/reviews/harness/harness-core/36264#-N96fZmGdq50lJ1qDKZm-r2-128:-N98Zqwv4_q7jPU5kYG9:b-mhu7ml) ([raw file](https://github.com/harness/harness-core/blob/7c495679054110d6c0cacc8524f06e7ccd7e08c6/400-rest/src/main/java/software/wings/sm/states/k8s/K8sDelete.java#L128)):*
> done
Done

**Review by shubham-in10se — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36264)* status: 2 change requests. 4. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java` line 171 at r3](https://reviewable.io/reviews/harness/harness-core/36264#-N9BtU9zAb2frbuNeWYq:-N9G2p0GAuQLh05Ubab5:b3ezyjc) ([raw file](https://github.com/harness/harness-core/blob/a085b968f96a6da4f8c4c1f7bc78ff7d569c349d/260-delegate/src/main/java/software/wings/delegatetasks/k8s/taskhandler/K8sDeleteTaskHandler.java#L171)):*
<details><summary><i>Previously, TarunUba1 (Tarun Uba) wrote…</i></summary><blockquote>

This case is already handled by existing method
</blockquote></details>

if `k8sDeleteTaskParameters.getResources()` is null, then `resourcesToDelete` will be null, and then in the subsequent if condition: `resourcesToDelete.contains()`, a NPE will be thrown.



_Code quote:_
```Java
k8sDeleteTaskParameters.getResources()
```
___
*[`930-delegate-tasks/src/main/java/software/wings/helpers/ext/k8s/request/K8sDeleteTaskParameters.java` line 73 at r4](https://reviewable.io/reviews/harness/harness-core/36264#-N9G3Jje8hUzs0nVgMfz:-N9G3Jje8hUzs0nVgMg-:b-2fo3qc) ([raw file](https://github.com/harness/harness-core/blob/f875bf0c42e0b78a1ec02ca2d66b68f4f62f6e7f/930-delegate-tasks/src/main/java/software/wings/helpers/ext/k8s/request/K8sDeleteTaskParameters.java#L73)):*
> ```Java
> 
>   public boolean isK8sCanaryDelete() {
>     return k8sPrevCanaryDeploy;
> ```

rename the member variable as well.



_Code quote:_
```Java
k8sPrevCanaryDeploy
```

**Review by shubham-in10se — DISMISSED** [review]
> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36264)* status: 1 change requests. 3. unresolved discussions.

**Review by abosii-harness — APPROVED** [review]
> Reviewed 5 of 8 files at r1, 1 of 4 files at r3, 3 of 5 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36264)* status: Some of the checks are too old. 1 change requests. 3. unresolved discussions.

**Review by shubham-in10se — APPROVED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36264)* status: Some of the checks are too old. 1 change requests. 3. unresolved discussions.

**Comment by TarunUba1** [comment]
trigger codebasehash

**Review by namantalaycha10 — APPROVED** [review]

**Review by namantalaycha10 — COMMENTED** [review]
> Reviewed 5 of 8 files at r1, 1 of 4 files at r3, 2 of 5 files at r4, 3 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36264)* status: Some of the checks are too old. 3. unresolved discussions.

**Comment by namantalaycha10** [comment]
please be sure these changes are tested with old delegate as well

**Review by TarunUba1 — COMMENTED** [review]
> Reviewed 5 of 8 files at r1, 1 of 4 files at r3, 2 of 5 files at r4, 3 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36264)* status: Some of the checks are too old. 3. unresolved discussions.

**Review by TarunUba1 — COMMENTED** [review]
Done

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36264#-:-NAIxMexDqraAjRSZZn5:b-896fjx)* status: Some of the checks are too old. 3. unresolved discussions.

**Review by TarunUba1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36264)* status: Some of the checks are too old. 3. unresolved discussions. 

___
*[`930-delegate-tasks/src/main/java/software/wings/helpers/ext/k8s/request/K8sDeleteTaskParameters.java` line 73 at r4](https://reviewable.io/reviews/harness/harness-core/36264#-N9G3Jje8hUzs0nVgMfz:-NAIyU03A5dLsKDVa6W5:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/f875bf0c42e0b78a1ec02ca2d66b68f4f62f6e7f/930-delegate-tasks/src/main/java/software/wings/helpers/ext/k8s/request/K8sDeleteTaskParameters.java#L73)):*
<details><summary><i>Previously, shubham-in10se (Shubham Maheshwari) wrote…</i></summary><blockquote>

rename the member variable as well.
</blockquote></details>

Done

**Review by namantalaycha10 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36264)* status: Some of the checks are too old.

**Review by TarunUba1 — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36264#-:-NAJ-tX33ZvR3IJhXzrn:bnfp4nl)* status: Some of the checks are too old.

**Pull request merged**